allennlp/modules/encoder_base.py

```diff
@@ -0,0 +1,268 @@
+        # calling self._module, then fill with zeros.
+
+        # First count how many sequences are empty.
+        batch_size, _ = mask.size()
+        num_valid = torch.sum(mask[:, 0]).int().data[0]
+
+        sequence_lengths = get_lengths_from_binary_sequence_mask(mask)
+        sorted_inputs, sorted_sequence_lengths, restoration_indices, sorting_indices =\
+            sort_batch_by_length(inputs, sequence_lengths)
+
+        # Now create a PackedSequence with only the non-empty, sorted sequences.
+        packed_sequence_input = pack_padded_sequence(sorted_inputs[:num_valid, :, :],
+                                                     sorted_sequence_lengths[:num_valid].data.tolist(),
+                                                     batch_first=True)
+        # Prepare the initial states.
+        if not self._stateful:
+            initial_states = hidden_state
+        else:
+            initial_states = self._get_initial_states(batch_size, num_valid, sorting_indices)
+
+        # Actually call the module on the sorted PackedSequence.
+        module_output, final_states = module(packed_sequence_input, initial_states)
+
+        return module_output, final_states, restoration_indices, num_valid
+
+    def _get_initial_states(self,
+                            batch_size: int,
+                            num_valid: int,
+                            sorting_indices: torch.LongTensor) -> Optional[RnnState]:
+        """
+        Parameters
+        ----------
+        batch_size : ``int``, required.
+            The batch size can change size across calls to stateful RNNs, so we need
+            to know if we need to expand or shrink the states before returning them.
+            Expanded states will be set to zero.
+        num_valid : ``int``, required.
+            The batch may contain completely padded sequences which get removed before
+            the sequence is passed through the encoder. We also need to clip these off
+            of the state too.
+        sorting_indices ``torch.LongTensor``, required.
+            Pytorch RNNs take sequences sorted by length. When we return the states to be
+            used for a given call to ``module.forward``, we need the states to match up to
+            the sorted sequences, so before returning them, we sort the states using the
+            same indices used to sort the sequences.
+
+        Returns
+        -------
+        This method has a complex return type because it has to deal with the first time it
+        is called, when it has no state, and the fact that types of RNN have heterogeneous
+        states.
+
+        If it is the first time the module has been called, it returns ``None``, regardless
+        of the type of the ``Module``.
+
+        Otherwise, for LSTMs, it returns a tuple of ``torch.Tensors`` with shape
+        ``(num_layers, num_valid, state_size)`` and ``(num_layers, num_valid, memory_size)``
```

Comment:
An explanation for these shapes is in order, I think - why is `num_layers` first?  This is particularly important because you can pass in an initial hidden state to the `forward` method.  I suppose adding this explanation above, where the docs for the hidden state were missing, would be enough.

Reply:
This is how all states are returned in pytorch